primitives/state-machine/src/testing.rs

```diff
@@ -135,21 +135,19 @@ impl<H: Hasher, N: ChangesTrieBlockNumber> TestExternalities<H, N>
 
 	/// Return a new backend with all pending value.
 	pub fn commit_all(&self) -> InMemoryBackend<H> {
-		let top: Vec<_> = self.overlay.committed.top.clone().into_iter()
-			.chain(self.overlay.prospective.top.clone().into_iter())
-			.map(|(k, v)| (k, v.value)).collect();
+		let top: Vec<_> = self.overlay.changes(None)
+			.map(|(k, v)| (k.clone(), v.value().cloned()))
+			.collect();
 		let mut transaction = vec![(None, top)];
 
-		self.overlay.committed.children_default.clone().into_iter()
-			.chain(self.overlay.prospective.children_default.clone().into_iter())
-			.for_each(|(_storage_key, (map, child_info))| {
-				transaction.push((
-					Some(child_info),
-					map.into_iter()
-						.map(|(k, v)| (k, v.value))
-						.collect::<Vec<_>>(),
-				))
-			});
+		for child_info in self.overlay.child_infos() {
```

Comment:
Here child_infos is currently building an unnecessary btreeset.
Also we are iterating child infos first in `child_infos` function, then we access the corresponding values a second time in `changes` when we could have use them from the first iteration.
Depending on what should be the next steps I wouldn't spend to much time on this (just avoid the internal btreeset).
Different api could be use, I just check I did use `chidren_iter_overlay` function in #3263, IIrc it is probably simply returning an double iterator.
But truely at this point we should not expect huge number of items in child trie.

Reply:
Yes this is doing work that is strictly not necessary by going over the children multiple times. However, as this is only test code I opted for not adding an additional API just for the sake of some potential speedup that does not even matter.

Reply:
Ah, we could think about making `child_infos` a test function then?

Reply:
No it is also used in production code.